### PR TITLE
Update ECPoint type

### PR DESCRIPTION
**Pull request by lulumel0n**
Per https://github.com/google/go-tpm/issues/115
Allow the ECCParams to supply buffers as X and Y for the Point field, in order to comply with the Default ECC EK Public Area Template.

When needed, the buffer will be converted to Int.

This change is not backward compatible, existing tests/code using ECPoint will need to be updated.

### CODE REVIEW

File: tpm2/structures.go [hunk 1]
```diff
@@ -256,23 +256,25 @@ type ECCParams struct {
...
 	}
-	return p.X
+	return x.SetBytes(p.XRaw)
 }
 
-func (p *ECPoint) y() *big.Int {
-	if p == nil || p.Y == nil {
-		return big.NewInt(0)
+func (p ECPoint) Y() *big.Int {
+	y := big.NewInt(0)
```
Comment (josephlr) [inline tpm2/structures.go]: you can just use `new(big.Int).SetBytes(x)` and `new(big.Int).SetBytes(y)` here. The nil slice of length zero works fine with `SetBytes`.

File: tpm2/structures.go [hunk 2]
```diff
@@ -295,8 +297,9 @@ func (p *ECCParams) encode() ([]byte, error) {
 	if err != nil {
 		return nil, fmt.Errorf("encoding KDF: %v", err)
 	}
-	x, y := p.Point.x().Bytes(), p.Point.y().Bytes()
-	point, err := tpmutil.Pack(tpmutil.U16Bytes(x), tpmutil.U16Bytes(y))
+
```
Comment (josephlr) [inline tpm2/structures.go]: Delete these blank lines.

File: tpm2/tpm2_test.go [hunk 3]
```diff
@@ -727,7 +727,7 @@ func TestSign(t *testing.T) {
 					Hash: AlgSHA256,
 				},
 				CurveID: CurveNISTP256,
-				Point:   ECPoint{X: big.NewInt(0), Y: big.NewInt(0)},
+				Point:   ECPoint{XRaw: big.NewInt(0).Bytes(), YRaw: big.NewInt(0).Bytes()},
```
Comment (josephlr) [inline tpm2/tpm2_test.go]: Just delete this line (the default value of `XRaw=nil, YRaw=nil` is fine here)

File: tpm2/encoding_test.go [hunk 4]
```diff
@@ -385,6 +385,14 @@ func TestECCParamsEncodeDecode(t *testing.T) {
 		t.Fatalf("decodeECCParams: %v", err)
 	}
 
+	if params.Point.X().Cmp(pk.PublicKey.X) != 0 || params.Point.Y().Cmp(pk.PublicKey.Y) != 0 {
+		t.Fatalf("Point deserialized: (%d, %d)\nGenerated: (%d, %d)", params.Point.X(), params.Point.Y(), pk.PublicKey.X, pk.PublicKey.Y)
```
Comment (josephlr) [inline tpm2/encoding_test.go]: For these `FatalF` messages, just use the shorter `"got: %+v\nwant: %+v"` format string, like we do below.

**Comment by josephlr** [comment]
Also, please update the PR description to reflect the new changes

File: tpm2/structures.go (hunk 1 above)
Comment (lulumel0n) [reply tpm2/structures.go]: Done

File: tpm2/structures.go (hunk 2 above)
Comment (lulumel0n) [reply tpm2/structures.go]: Done

File: tpm2/tpm2_test.go (hunk 3 above)
Comment (lulumel0n) [reply tpm2/tpm2_test.go]: Done

File: tpm2/encoding_test.go (hunk 4 above)
Comment (lulumel0n) [reply tpm2/encoding_test.go]: Done

File: tpm2/encoding_test.go [hunk 5]
```diff
@@ -385,6 +385,19 @@ func TestECCParamsEncodeDecode(t *testing.T) {
 		t.Fatalf("decodeECCParams: %v", err)
 	}
 
+	if params.Point.X().Cmp(pk.PublicKey.X) != 0 {
+		t.Fatalf("got: %+v\nwant: %+v", params.Point.X(), pk.PublicKey.X)
```
Comment (awly) [inline tpm2/encoding_test.go]: Clarify what got/want in these failure messages are

Comment (awly) [reply tpm2/encoding_test.go]: Sorry, I meant in the actual `t.Fatalf` text, so you don't need to look at test code to understand which check failed.

Comment (lulumel0n) [reply tpm2/encoding_test.go]: np, added

File: tpm2/structures.go [hunk 6]
```diff
@@ -295,8 +289,7 @@ func (p *ECCParams) encode() ([]byte, error) {
 	if err != nil {
 		return nil, fmt.Errorf("encoding KDF: %v", err)
 	}
-	x, y := p.Point.x().Bytes(), p.Point.y().Bytes()
-	point, err := tpmutil.Pack(tpmutil.U16Bytes(x), tpmutil.U16Bytes(y))
+	point, err := tpmutil.Pack(tpmutil.U16Bytes(p.Point.XRaw), tpmutil.U16Bytes(p.Point.YRaw))
```
Comment (josephlr) [inline tpm2/structures.go]: `tpmutil.U16Bytes` isn't needed here (as that's already the type of `p.Point.XRaw`.

`tpmutil.Pack(p.Point.XRaw, p.Point.YRaw)` should work.

**Comment by josephlr** [comment]
@awly, this is a backwards incompatible change, but makes dealing with encoding/decoding issues much easier.

Would you want to do the same thing for `RSAParams.Modulus` (in a followup PR)? The main downside is that RSA keys are more common, so it might break a little more code, but I think we should do it.

File: tpm2/structures.go (hunk 6 above)
Comment (lulumel0n) [reply tpm2/structures.go]: good catch, done

**Comment by awly** [comment]
SGTM, we always assume that importers will lock the version of this library.
Breaking API isn't a problem.

**Pull request merged**